// vue-auto-routing/index.js
function index() {
  return import(/* webpackChunkName: "page-index" */ '@/views/pages/index.vue')
}
function about() {
  return import(/* webpackChunkName: "page-about" */ '@/views/pages/about.vue')
}
function faq() {
  return import(/* webpackChunkName: "page-faq" */ '@/views/pages/faq.vue')
}
function store_index() {
  return import(
    /* webpackChunkName: "page-store-index" */ '@/views/pages/store/index.vue'
  )
}
function store_location_index() {
  return import(
    /* webpackChunkName: "page-store-location-index" */ '@/views/pages/store/location/index.vue'
  )
}
function store__id_index() {
  return import(
    /* webpackChunkName: "page-store-id-index" */ '@/views/pages/store/_id/index.vue'
  )
}

export default [
  {
    name: 'index',
    path: '',
    component: index,
  },
  {
    name: 'about',
    path: 'about',
    component: about,
  },
  {
    name: 'faq',
    path: 'faq',
    component: faq,
  },
  {
    name: 'store',
    path: 'store',
    component: store_index,
    meta: {
      requiresAuth: false,
    },
  },
  {
    name: 'store-location',
    path: 'store/location',
    component: store_location_index,
  },
  {
    name: 'store-id',
    path: 'store/:id',
    component: store__id_index,
  },
]
